### PR TITLE
Prototype system initialization changes for comment

### PR DESCRIPTION
**Pull request by 0xTJ**
This is my prototype for comment, as discussed in #303. I've cut back the scope of my changes a bit.

In the end, most of my changes ended up preserving the original code, just reorganizing it. I've split out some code from `bootstrap.asm`, so that file is a good bit simpler. It has mostly ended up in the revision- or UART-specific files. While some of this technically repeats some bits of codes within routines, a lot of them were a bit hard to read with the amount of `ifd`s in them.

One major change is that the EFPT is completely initialized with dummy pointers before calling any system init functions, to make it easier to handle no default console being present.

I've tried to make the serial chips' code a little bit more orthogonal to the board revisions, though they're still tightly coupled.

I had to add `-fno-delete-null-pointer-checks` to the `Makefile`, because GCC would generate a `trap #7` when it could prove that a certain access would always dereference `NULL` (this happened when I changed a global pointer with address 0 to be `const`).

I added a `hot_boot` routine, which is like `warm_boot`, but doesn't print the banner. It's called from `main1` to make the state consistent with a warm boot when calling the program loader.

Actually running it, the changes to code flow are minor, and I don't think there would be any noticeable changes (at least in theory, practice might be different).

I haven't checked that it runs on r1.x hardware (I currently only have one MC68010, and it's a pain of a chip to transplant), just that it builds for `REVISION1X=true HUGEROM=false`.

I'd appreciate any comments on this. I've tried to be good about committing per-change, so it shouldn't be too hard to undo specific changes.

This is inspired by #303, but that proposed PLD change would close that issue.

### CODE REVIEW

**Comment by roscopeco** [comment]
Apologies for the delay on this @0xTJ , I have been a bit short on time and want to pull the changes and test / look at them locally. I should be able to do that tomorrow 😃

File: code/firmware/rosco_m68k_v2.1/bootstrap.asm [hunk 1]
```diff
@@ -61,51 +60,55 @@ VECTORS:
...
+    ; Copy exception vectors table to RAM at VECTORS_LOAD (0x00000000).
+    ; VBR defaults to that location anyway for 68000 compatibility.
+    lea     (VECTORS),A0                ; Vectors in ROM into A0 (source)
+    lea     (VECTORS_LOAD),A1           ; Vectors in RAM into A1 (destination)
+    move.l  #VECTORS_COUNT,D0           ; Count into D0 (DBcc only uses word size)
+    bra     .ISR_COPY_START             ; Jump to DBcc to start loop
+.ISR_COPY_LOOP:
+    move.l  (A0)+,(A1)+                 ; Copy long source to dest, with postincrement.
+.ISR_COPY_START:
+    dbf     D0,.ISR_COPY_LOOP           ; Decrement D0 and loop if not negative, ignore cc
```
Comment (roscopeco) [inline code/firmware/rosco_m68k_v2.1/bootstrap.asm:82]: Nice, much better 👍

File: code/firmware/rosco_m68k_v2.1/main1.c [hunk 2]
```diff
@@ -71,10 +72,10 @@ noreturn void main1();
 
 noreturn void linit() {
   // copy .data
-  for (uint32_t *dst = &_data_start, *src = &_code_end; dst < &_data_end; *dst = *src, dst++, src++);
+  for (uint16_t *dst = &_data_start, *src = &_code_end; dst < &_data_end; *dst = *src, dst++, src++);
```
Comment (roscopeco) [inline code/firmware/rosco_m68k_v2.1/main1.c:75]: Interesting. Why this change?

File: code/firmware/rosco_m68k_v2.1/rosco_m68k_private.asm [hunk 3]
```diff
@@ -20,12 +20,12 @@
 ; bit 14    : Requires larger system data area
 ; bit 15    : Snapshot version
   ifd HUGEROM
-RELEASE_VER     equ     $E0000220
+RELEASE_VER     equ     $E0000230
```
Comment (roscopeco) [inline code/firmware/rosco_m68k_v2.1/rosco_m68k_private.asm:23]: Did you mean to bump this? 2.20 is still in DEV, AFAIK this change should be part of that eventual release...

**Review by roscopeco — COMMENTED** [review]
This looks generally good, and is a significant win in terms of clean separation and tidying up a bunch of things. It tests fine here on r2 boards in various HW configurations. I too have not had chance to test on an r1, but will do that next.

File: code/firmware/rosco_m68k_v2.1/rosco_m68k_private.asm (hunk 3 above)
Comment (0xTJ) [reply code/firmware/rosco_m68k_v2.1/rosco_m68k_private.asm]: Very good point, I've reverted that.

File: code/firmware/rosco_m68k_v2.1/main1.c (hunk 2 above)
Comment (roscopeco) [reply code/firmware/rosco_m68k_v2.1/main1.c:75]: Okay, yep, good points (and well spotted!). Given that I’d go with keeping this change 👍

**Review by roscopeco — APPROVED** [review]
Lightly tested on r1.23 board with AT28C256 ROMs (built with `REVISION1X=true`) and it works great. As mentioned, already tested on r2.x boards in various configurations.

Overall, I think this is a great change, much cleaner. Thanks again for the good work @0xTJ !

**Comment by 0xTJ** [comment]
Thanks for the review @roscopeco!
